playwright/e2e/settings/encryption-user-tab/index.ts

```diff
@@ -0,0 +1,118 @@
+        await this.enterRecoveryKey(recoveryKey);
+        await this.page.getByRole("button", { name: "Done" }).click();
+    }
+
+    /**
+     * Fill the recovery key in the dialog
+     * @param recoveryKey
+     */
+    async enterRecoveryKey(recoveryKey: GeneratedSecretStorageKey) {
+        // Select to use recovery key
+        await this.page.getByRole("button", { name: "use your Security Key" }).click();
+
+        // Fill the recovery key
+        const dialog = this.page.locator(".mx_Dialog");
+        await dialog.getByRole("textbox").fill(recoveryKey.encodedPrivateKey);
+        await dialog.getByRole("button", { name: "Continue" }).click();
+    }
+
+    /**
+     * Get the encryption tab content
+     */
+    getEncryptionTabContent() {
+        return this.page.getByTestId("encryptionTab");
+    }
+
+    /**
+     * Delete the key backup for the given version
+     * @param backupVersion
+     */
+    async deleteKeyBackup(backupVersion: string) {
+        const client = await this.app.client.prepareClient();
+        await client.evaluate(async (client, backupVersion) => {
+            await client.getCrypto()?.deleteKeyBackupVersion(backupVersion);
+        }, backupVersion);
+    }
+
+    /**
+     * Get the security key from the clipboard and fill in the input field
+     * Then click on the finish button
+     * @param screenshot
+     */
+    async confirmRecoveryKey(screenshot: `${string}.png`) {
+        const dialog = this.getEncryptionTabContent();
+        await expect(dialog.getByText("Enter your recovery key to confirm")).toBeVisible();
+        await expect(dialog).toMatchScreenshot(screenshot);
+
+        const handle = await this.page.evaluateHandle(() => navigator.clipboard.readText());
+        const clipboardContent = await handle.jsonValue();
+        await dialog.getByRole("textbox").fill(clipboardContent);
+        await dialog.getByRole("button", { name: "Finish set up" }).click();
+        await expect(dialog).toMatchScreenshot("default-recovery.png");
+    }
+
+    /**
+     * Remove the cached secrets from the indexedDB
+     */
+    async deleteCachedSecrets() {
+        await this.page.evaluate(async () => {
+            const removeCachedSecrets = new Promise((resolve) => {
+                const request = window.indexedDB.open("matrix-js-sdk::matrix-sdk-crypto");
+                request.onsuccess = async (event: Event & { target: { result: IDBDatabase } }) => {
+                    const db = event.target.result;
+                    const request = db.transaction("core", "readwrite").objectStore("core").delete("private_identity");
+                    request.onsuccess = () => {
+                        db.close();
+                        resolve(undefined);
+                    };
+                };
+            });
+            await removeCachedSecrets;
+        });
+        await this.page.reload();
+    }
```

Comment:
Given this is only used once, and it is a rather ... niche ... usecase, I'm not sure it is a good fit for this playwright fixture.

I'd be inclined to make it a private function in `recovery.spec.ts`, which takes a `page` param.

Reply:
I also think its worth writing a comment about why we are digging about in the guts of indexeddb here. Clearly this is not a thing that a real user will ever do, so what scenario are we trying to simulate here? What set of circumstances could lead to the identity existing in 4S but it not being cached?